sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/CombineFns.java

```diff
@@ -1038,4 +1034,43 @@ public void verifyDeterministic() throws NonDeterministicException {
         "Cannot compose with tuple tag %s because it is already present in the composition.",
         outputTag);
   }
+
+  /**
+   * Populate display data for the {@code combineFns} that make up a composed combine transform.
+   *
+   * <p>The same combineFn class may be used multiple times, in which case we must take special care
+   * to register display data with unique namespaces.
+   */
+  private static void populateDisplayData(
+      DisplayData.Builder builder, List<? extends HasDisplayData> combineFns) {
+
+    // NB: ArrayListMultimap necessary to maintain ordering of combineFns of the same type.
+    Multimap<Class<?>, HasDisplayData> combineFnMap = ArrayListMultimap.create();
+
+    for (int i = 0; i < combineFns.size(); i++) {
+      HasDisplayData combineFn = combineFns.get(i);
+      builder.add("combineFn" + (i + 1), combineFn.getClass());
+      combineFnMap.put(combineFn.getClass(), combineFn);
+    }
+
+    for (Map.Entry<Class<?>, Collection<HasDisplayData>> combineFnEntries :
+        combineFnMap.asMap().entrySet()) {
+
+      Collection<HasDisplayData> classCombineFns = combineFnEntries.getValue();
+      if (classCombineFns.size() == 1) {
+        // Only one combineFn of this type, include it directly.
+        builder.include(Iterables.getOnlyElement(classCombineFns));
+
+      } else {
+        // Multiple combineFns of same type, add a namespace suffix so display data is
+        // unique and ordered.
+        String baseNamespace = combineFnEntries.getKey().getName();
+        for (int i = 0; i < combineFns.size(); i++) {
+          HasDisplayData combineFn = combineFns.get(i);
+          String namespace = String.format("%s$%d", baseNamespace, i + 1);
```

Comment:
Not sure it matters, but we may want to use something other than `$`, since that is already used for anonymous inner classes. Maybe `#`? Although at that point, we've already gone to the point of having the `path` (in this case the path is just a number).